SWAPI Client Creation and List Implementation

eddiescj opened the pull request:
The main target of this PR is to make sure there is a List All endpoint provided to api consumers where they can see all planets registered.

Also, is important make sure the code is well done, tests are running and the default structure is complete.

Commit:
feature: Adding SWAPI Planet struct to
structs domain will bem used to map the
response from Star wars api.

Commit:
refactor: changes description in line below

- Changing path structure to have lesser folders
- Adding new enviroment variables

Commit:
feature: client swapi client where we will get our planets as second option

Commit:
refactor: changing package structure

Commit:
refactor: details in lines below
- Refactoring domain to be isolated
- Creating service and repo operations

Commit:
refactor: fixing warnings

Commit:
refactor: solving lint issues

EddieSCJ commented:
@dubonzi da uma olhada dps, ainda n ta pronto mas termino essa semana essa primeira parte

dubonzi commented on app/planets/list/gateway/swapi_gateway.go:
```diff
@@ -0,0 +1,48 @@
+	baseURL string
+}
+
+func NewSWAPIGateway() SWAPIGateway {
```
Ao invés de retornar a interface, é uma boa retornar o tipo concreto e deixar a definição da interface do lade de quem vai consumir, pra reduzir o acoplamento. Isso vale para os outros locais onde uma interface tá sendo retornada também, Service por exemplo.

Não tem problema em retornar o `swapiGateway` privado no construtor. O que isso ta fazendo é somente impedir que outro pacote crie um `swapiGateway` sem usar o construtor. 

https://bryanftan.medium.com/accept-interfaces-return-structs-in-go-d4cab29a301b

eddiescj replied:
Simplesmente genial esse artigo.

dubonzi commented on app/planets/list/gateway/swapi_gateway.go:
```diff
@@ -0,0 +1,48 @@
+	url := s.baseURL + "/planets"
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return nil, err
```
Eu adicionaria um pouco mais de contexto nesses erros pra facilitar o entendimento e melhorar os logs. Da pra usar o método `Wrap` do pacote `"github.com/pkg/errors"`.

ex: `return errors.Wrap(err, "error building swapi request")`

Esse método cria um 'stack' dos erros e quando você printa ele sai printando todos os erros e mensagens que você adicionou. Os métodos desse pacote também adicionam stack trace aos erros, dá uma lida na documentação.

dubonzi commented on app/planets/list/gateway/swapi_gateway.go:
```diff
@@ -0,0 +1,48 @@
+	return gateway
+}
+
+func (s swapiGateway) GetPlanets(filter Filter) (*http.Response, error) {
```
Eu notei que você não ta fazendo uso do `context.Context` nos métodos. É meio que um 'standard' receber um `context.Context` e ir propagando ele pela aplicação. Isso é bem útil por exemplo quando você tem integração com algum APM tipo o datadog.

Como é uma API web, cada request que chega no handler/controller possui o seu próprio context, você pode propagar esse mesmo context até aqui e usar ele como context do request. 

```go
func GetPlanets(ctx context.Context, filter Filter) (*http.Response, error) {
    ...
    req := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
    s.Client.Do(req)
    ...
}
```

Se tiver usando integração com o Datadog por ex, o middleware do datadog vai interceptar essa chamada e olhar pro context pra extrair as infos e montar os traces e span.

https://github.com/DataDog/dd-trace-go/blob/e7717adc3b86ceecc95702ad007ea3115d442fa3/contrib/net/http/roundtripper.go#L42

Isso tudo também vale pro mongo (vi também que não passou o context la) e pra qualquer outra parte da aplicação onde você veja utilidade para o context. Vale lembrar que o context também pode ser cancelado, seja via timeout ou chamando o método cancel (`context.WithCancel`), então vale dar uma lida sobre pra descobrir os usos dele.

eddiescj replied:
Dei uma olhada, me passei nisso, mas e tipo o mdc do spring

dubonzi commented on app/planets/list/storage/mongo.go:
```diff
@@ -0,0 +1,47 @@
+}
+
+func (r repository) GetAll(mongoOptions MongoOptions) ([]model.PlanetMongo, error) {
+	findOptions := options.Find().
```
Da pra criar um método tipo `Build()` na struct `MongoOptions` e retornar o `FindOptions` construído, iria deixar esse método aqui mais enxuto.

dubonzi commented on app/planets/list/storage/mongo.go:
```diff
@@ -0,0 +1,47 @@
+	}
+
+	var planets []model.PlanetMongo
+	if err := cursor.All(nil, &planets); err != nil {
```
Mesma questão do context, nesse caso aqui você ta passando nil, o que não é muito recomendado, no mínimo passa o `context.TODO()` que é um context vazio.

dubonzi commented on app/planets/model/mongo.go:
```diff
@@ -0,0 +1,38 @@
+}
+
+func ToDomainList(mongoPlanets []PlanetMongo) []Planet {
+	var domainList []Planet
```
Dica de otimização: quando você sabe o tamanho final do slice, já cria ele com esse tamanho, dependendo da quantidade de dados, você salva bastante cpu/memória.

Nesse caso aqui ficaria `domainList := make([]Planet, 0, len(mongoPlanets))`

Escrevi um [post](https://blog.eduardobonzi.dev/improving-performance-with-slices-in-go) sobre isso se quiser dar uma lida ;)

dubonzi commented on app/planets/model/swapi.go:
```diff
@@ -0,0 +1,37 @@
+}
+
+func (r SWAPIResponseBody) ToDomainList() []Planet {
+	var planets []Planet
```
Mesma dica de otimização.

EddieSCJ commented:
@dubonzi opa, não consegui terminar o código que estava fazendo antes do fim de semana pois precisei castrar minhas pets e acabei ficando sobrecarregado.

Porém li as dicas aqui e tu acertou em cheio as partes que já estavam finalizadas, curti demais as dicas e certamente vou trabalhar em cima delas ASAP, mt obrigado <3

Commit:
refactor: adjusting code to return struct and receive interfaces

Commit:
refactor: clean up
- Wrapping better errors
- Improving slices performance
- Improving logs
- Setting context

Commit:
refactor: clean up
- adjusting logs
- breaking code blocks

Commit:
refactor: goimport files

Commit:
refactor: creating tests to gateway layer

Commit:
refactor: creating some unit tests

Commit:
refactor: adding tests and new mock

Commit:
refactor: addjusting lints

Commit:
refactor: adjusting lints

Commit:
refactor: adding mongo integrated tests

Commit:
refactor: adding mongo integrated tests

Commit:
refactor: adding tests to external api gateway

Commit:
refactor: adding unit tests to handler

Commit:
refactor: setting logs

Commit:
refactor: fixing lint

Commit:
refactor: adding route

Commit:
refactor: fix lint

Commit:
refactor: adjusting logs

Commit:
refactor: see description below
- In order to have the latest version of dockertest, it was updated to v3

Commit:
refactor: see description below
- Creating dockerfile

Commit:
refactor: see description below
- Added a new test in options unit test
- Adjust MongoOptions build condition, once it was failing in the last condition

Commit:
refactor: see description below
- Created new builder and divided it into two pieces, one to build the operation handler, another to bind context with echo
- Created integration tests to list operation

Commit:
refactor: fixing import issue

Commit:
refactor: adjusting code smells
- Seems like I was calling same endpoint many times, so, make sense define a constant for them

Commit:
refactor: adjusting code smells

EddieSCJ commented:
@dubonzi @jpfaria @jeancabral @fernandomalmeida

your time to review guys
